File: src/backend/microservices/visualization/src/main/java/org/phyloviz/pwp/visualization/service/VisualizationService.java
package org.phyloviz.pwp.visualization.service;

import org.phyloviz.pwp.shared.service.dtos.files.isolate_data.GetIsolateDataRowsOutput;
import org.phyloviz.pwp.shared.service.dtos.files.typing_data.GetTypingDataProfilesOutput;
import org.phyloviz.pwp.shared.service.dtos.files.typing_data.GetTypingDataSchemaOutput;
import org.phyloviz.pwp.shared.service.dtos.tree_view.GetTreeViewOutput;
import org.phyloviz.pwp.shared.service.dtos.tree_view.SaveTreeViewInput;
import org.phyloviz.pwp.shared.service.dtos.tree_view.SaveTreeViewOutput;

import java.util.List;

/**
 * Service for the Visualization Microservice.
 */
public interface VisualizationService {

    /**
     * Gets a distance matrix.
     *
     * @param projectId        the id of the project
     * @param datasetId        the id of the dataset
     * @param distanceMatrixId the id of the distance matrix
     * @param userId           the id of the user
     * @return the distance matrix in string
     */
    String getDistanceMatrix(String projectId, String datasetId, String distanceMatrixId, String userId);

    /**
     * Gets a tree.
     *
     * @param projectId the id of the project
     * @param datasetId the id of the dataset
     * @param treeId    the id of the tree
     * @param userId    the id of the user
     * @return the tree in newick string format
     */
    String getTree(String projectId, String datasetId, String treeId, String userId);

    /**
     * Gets a tree view.
     *
     * @param projectId  the id of the project
     * @param datasetId  the id of the dataset
     * @param treeViewId the id of the tree view
     * @param userId     the id of the user
     * @return the tree view
     */
    GetTreeViewOutput getTreeView(String projectId, String datasetId, String treeViewId, String userId);

    /**
     * Saves a tree view coordinates and applied transformations.
     *
     * @param projectId  the id of the project
     * @param datasetId  the id of the dataset
     * @param treeViewId the id of the tree view
     * @param inputModel the input model
     * @param userId     the id of the user
     * @return the tree view
     */
    SaveTreeViewOutput saveTreeView(String projectId, String datasetId, String treeViewId, SaveTreeViewInput inputModel,
                                    String userId);

    /**
     * Gets a typing data's schema.
     *
     * @param projectId    the id of the project
     * @param typingDataId the id of the typing data
     * @param userId       the id of the user
     * @return the typing data schema
     */
    GetTypingDataSchemaOutput getTypingDataSchema(String projectId, String typingDataId, String userId);

    /**
     * Gets a typing data's profiles, with pagination.
     *
     * @param projectId    the id of the project
     * @param typingDataId the id of the typing data
     * @param limit        the limit of the profiles
     * @param offset       the offset of the profiles
     * @param userId       the id of the user
     * @return the typing data profiles
     */
    GetTypingDataProfilesOutput getTypingDataProfiles(String projectId, String typingDataId, int limit, int offset,
                                                      String userId);

    /**
     * Gets the keys of an isolate data.
     *
     * @param projectId     the id of the project
     * @param isolateDataId the id of the isolate data
     * @param userId        the id of the user
     * @return the isolate data keys
     */
    List<String> getIsolateDataKeys(String projectId, String isolateDataId, String userId);

    /**
     * Gets the rows of an isolate data, with pagination.
     *
     * @param projectId     the id of the project
     * @param isolateDataId the id of the isolate data
     * @param limit         the limit of the rows
     * @param offset        the offset of the rows
     * @param userId        the id of the user
     * @return the isolate data rows
     */
    GetIsolateDataRowsOutput getIsolateDataRows(String projectId, String isolateDataId, int limit, int offset,
                                                String userId);
}

File: src/frontend/src/Services/visualization/VisualizationService.ts
import {get} from "../utils/apiFetch"
import {WebApiUris} from "../WebApiUris"
import {GetTypingDataSchemaOutputModel} from "./models/getTypingDataSchema/GetTypingDataSchemaOutputModel"
import {GetTypingDataProfilesOutputModel} from "./models/getTypingDataProfiles/GetTypingDataProfilesOutputModel"
import {GetTypingDataFileOutputModel} from "./models/getTypingDataFile/GetTypingDataFileOutputModel"
import {GetIsolateDataSchemaOutputModel} from "./models/getIsolateDataSchema/GetIsolateDataSchemaOutputModel"
import {GetIsolateDataRowsOutputModel} from "./models/getIsolateDataProfiles/GetIsolateDataRowsOutputModel"
import {GetIsolateDataFileOutputModel} from "./models/getIsolateDataFile/GetIsolateDataFileOutputModel"
import {GetDistanceMatrixOutputModel} from "./models/getDistanceMatrix/GetDistanceMatrixOutputModel"
import {GetTreeOutputModel} from "./models/getTree/GetTreeOutputModel"
import {GetTreeViewOutputModel} from "./models/getTreeView/GetTreeViewOutputModel"
import {MockVisualizationService} from "./MockVisualizationService"


namespace VisualizationService {

    /**
     * Get the typing data schema.
     *
     * @param projectId The project id.
     * @param typingDataId The typing data id.
     * @returns The typing data schema.
     */
    export async function getTypingDataSchema(
        projectId: string,
        typingDataId: string
    ): Promise<GetTypingDataSchemaOutputModel> {
        return await get<GetTypingDataSchemaOutputModel>(WebApiUris.getTypingDataSchema(projectId, typingDataId))
    }

    /**
     * Get the typing data profiles.
     *
     * @param projectId The project id.
     * @param typingDataId The typing data id.
     * @returns The typing data profiles.
     */
    export async function getTypingDataProfiles(
        projectId: string,
        typingDataId: string
    ): Promise<GetTypingDataProfilesOutputModel> {
        return await get<GetTypingDataProfilesOutputModel>(WebApiUris.getTypingDataProfiles(projectId, typingDataId))
    }

    /**
     * Get the typing data file.
     *
     * @param projectId The project id.
     * @param typingDataId The typing data id.
     * @returns The typing data file.
     */
    export async function getTypingDataFile(
        projectId: string,
        typingDataId: string
    ): Promise<GetTypingDataFileOutputModel> {
        return await get<GetTypingDataFileOutputModel>(WebApiUris.getTypingDataFile(projectId, typingDataId))
    }

    /**
     * Get the isolate data schema.
     *
     * @param projectId The project id.
     * @param isolateDataId The isolate data id.
     * @returns The isolate data schema.
     */
    export async function getIsolateDataSchema(
        projectId: string,
        isolateDataId: string
    ): Promise<GetIsolateDataSchemaOutputModel> {
        return await get<GetIsolateDataSchemaOutputModel>(WebApiUris.getIsolateDataSchema(projectId, isolateDataId))
    }

    /**
     * Get the isolate data rows.
     *
     * @param projectId The project id.
     * @param isolateDataId The isolate data id.
     * @returns The isolate data rows.
     */
    export async function getIsolateDataRows(
        projectId: string,
        isolateDataId: string
    ): Promise<GetIsolateDataRowsOutputModel> {
        return await get<GetIsolateDataRowsOutputModel>(WebApiUris.getIsolateDataRows(projectId, isolateDataId))
    }

    /**
     * Get the isolate data file.
     *
     * @param projectId The project id.
     * @param isolateDataId The isolate data id.
     * @returns The isolate data file.
     */
    export async function getIsolateDataFile(
        projectId: string,
        isolateDataId: string
    ): Promise<GetIsolateDataFileOutputModel> {
        return await get<GetIsolateDataFileOutputModel>(WebApiUris.getIsolateDataFile(projectId, isolateDataId))
    }

    /**
     * Get the distance matrix.
     *
     * @param projectId The project id.
     * @param datasetId The dataset id.
     * @param distanceMatrixId The distance matrix id.
     * @returns The distance matrix.
     */
    export async function getDistanceMatrix(
        projectId: string,
        datasetId: string,
        distanceMatrixId: string
    ): Promise<GetDistanceMatrixOutputModel> {
        return await get<any>(WebApiUris.getDistanceMatrix(projectId, datasetId, distanceMatrixId))
    }

    /**
     * Get the tree.
     *
     * @param projectId The project id.
     * @param datasetId The dataset id.
     * @param treeId The tree id.
     * @returns The tree.
     */
    export async function getTree(
        projectId: string,
        datasetId: string,
        treeId: string
    ): Promise<GetTreeOutputModel> {
        return await get<any>(WebApiUris.getTree(projectId, datasetId, treeId))
    }

    /**
     * Get the tree view.
     *
     * @param projectId The project id.
     * @param datasetId The dataset id.
     * @param treeViewId The tree view id.
     * @returns The tree view.
     */
    export async function getTreeView(
        projectId: string,
        datasetId: string,
        treeViewId: string
    ): Promise<GetTreeViewOutputModel> {
        return await get<any>(WebApiUris.getTreeView(projectId, datasetId, treeViewId))
    }
}

const env = process.env.NODE_ENV

export default env === "development" ? MockVisualizationService : VisualizationService
